### PR TITLE
add valiation of staged.dependencies yaml output

### LINKED ISSUE #26: yaml validator for staged_dependencies

### PR DESCRIPTION
**Pull request by nikolas-burkoff**
Closes #26 

So I couldn't find a nice package to do this for us like [cerberus](https://github.com/pyeve/cerberus) - does anyone know one?

But as our schema is very simple so I've written a simple validator which allows a little bit of freedom if our specs change - without needing a rewrite - but is not a full parser (which would be overkill here) 

Also added a huge pile of tests to check it's working as expected - yay TDD.

### CODE REVIEW

File: tests/testthat/test-validate_yaml.R [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    - not_repo: test3/test
+      host: https://github.com
+  current_repo:
+    repo: test/test
+    host: https://github.com
+  "
+  expect_error(validate_staged_deps_yaml(yaml::read_yaml(text = text)))
+})
+
+test_that("staged dep yaml with array of current_repo is invalid", {
```
Comment (maximilianmordig) [inline tests/testthat/test-validate_yaml.R:167]: this seems to be almost the same as the test starting on line 81.

File: tests/testthat/test-validate_yaml.R [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  expect_silent(validate_staged_deps_yaml(yaml::read_yaml(text = text)))
+})
+
+#---- Invalid yaml ----
+
+test_that("empty staged dep yaml file throws error", {
+
+  text <- "---
+  "
+  expect_error(validate_staged_deps_yaml(yaml::read_yaml(text = text)))
```
Comment (maximilianmordig) [inline tests/testthat/test-validate_yaml.R:44]: You may test with a regex in `expect_error` here and below to make sure it throws the right kind of error.

File: R/utils.R [hunk 3]
```diff
@@ -69,3 +69,70 @@ check_dir_exists <- function(direc, prefix = "") {
...
+
+
+  # first check the required fields exist
+  required_fields <- lapply(required_schema, "[[", "name")
+  if (!all(required_fields %in% names(content))) {
+    stop("File ", file_name , " invalid, it must contain fields ", toString(required_fields))
+  }
+
+  # next check the contents of the fields is as expected
+  lapply(required_schema, function(field, content){
```
Comment (maximilianmordig) [inline R/utils.R:116]: you could omit passing `content` here

File: R/utils.R [hunk 4]
```diff
@@ -69,3 +69,70 @@ check_dir_exists <- function(direc, prefix = "") {
...
+  required_schema <- list(
+    list(name = "upstream_repos", content = c("repo", "host"), nullable = TRUE, array = TRUE),
+    list(name = "downstream_repos", content = c("repo", "host"), nullable = TRUE, array = TRUE),
+    list(name = "current_repo", content = c("repo", "host"), nullable = FALSE, array = FALSE)
+  )
+
+  # helper function to validate field values
+  check_single_entry <- function(content, expected_content, field_name) {
+    # check the required contents exist
+    if (!(all(expected_content %in% names(content)))) {
```
Comment (maximilianmordig) [inline R/utils.R:93]: I would rename `expected_content` to `expected_fields`.

File: R/utils.R [hunk 5]
```diff
@@ -69,3 +69,70 @@ check_dir_exists <- function(direc, prefix = "") {
...
+validate_staged_deps_yaml <- function(content, file_name = "") {
+
+  # A simplified schema object to capture the schema for the yaml file
+  # each entry of the list contains the top level field, with their name
+  # whether they can be NULL and their content. If array is TRUE then each element
+  # of the field should contain the listed contents (as characters)
+  # More complex schemas cannot yet be validated
+  # Fields in addition to these are ignored and no error is shown
+  required_schema <- list(
+    list(name = "upstream_repos", content = c("repo", "host"), nullable = TRUE, array = TRUE),
```
Comment (maximilianmordig) [inline R/utils.R:85]: would rename `content` to `subfields` or similar.

**Review by maximilianmordig — APPROVED** [review]
Good after addressing the minor comments above.

File: tests/testthat/test-validate_yaml.R (hunk 1 above)
Comment (nikolas-burkoff) [reply tests/testthat/test-validate_yaml.R:167]: good spot - removing this

**Pull request merged**